client/components/select/select.directive.js

```diff
@@ -0,0 +1,240 @@
+  extract: '@',       // Name of the property to extract and use in the select field (Ex: id)
+  allowClear: '=?',   // Enable the possibility to remove a selection
+  labelNew: '@',      // The label text for the creation of a new item
+  defaultNewItem: '@' // The default text value for the creation of a new item
+};
+
+/**
+ * Define the standard default values for the select directive
+ */
+var defaults = {
+  showProperty: 'name', // Attribute name to use to show the item in the select field
+  allowClear: true,     // It is allowed to unselect items
+  multiple: false,      // She select fields are single selection
+  noLabel: false,       // The label is visible
+  extract: 'id'         // Use the ID to identify items
+};
+
+/**
+ * Create standard select directive
+ *
+ * @param controllerName The controller name to manage the directive
+ * @param attributes The attributes to configure the directive
+ * @returns The directive
+ */
+function createSelectDirective(controllerName, attributes) {
+  return {
+    restrict: 'E',
+    controller: controllerName,
+    templateUrl: '/templates/components/select/select.template.html',
+    scope: attributes
+  };
+}
+
+/**
+ * Configure a controller for a select directive
+ *
+ * @param $scope The scope
+ * @param api The API service
+ * @param options The options for the controller configuration
+ */
+function setupController($scope, api, options) {
+  if (!$scope.prefix) {
+    throw new Error("The prefix attribute on select directive is not set.");
+  }
+
+  if (!options) {
+    throw new Error("No options defined to setup the select controller.");
+  }
+
+  if (!options.defaults) {
+    throw new Error("No defaults provided to setup the select controller.");
+  }
+
+  if (!options.fetchUrl) {
+    throw new Error("The fetchUrl must be provided.");
+  }
+
+  // Setup the default options and values for the controller
+  _.defaults($scope, _.extend({
+    config: {
+      newItem: false
+    },
+    choices: []
+  }, _.defaults(options.defaults, defaults)));
+
+  $scope.$watch('config.newItem', function(value) {
+    if (value) {
+      var previousItem = $scope.modelObject[$scope.modelProperty];
+      // create a new object if a new item is to be created
+      $scope.modelObject[$scope.modelProperty] = {};
+      // pre-fill it with either the previously selected item, or with a default new item value
+      $scope.modelObject[$scope.modelProperty] = previousItem ? previousItem : $scope.defaultNewItem;
+    } else if (value === false && $scope.choices.length && $scope.autoSelect) {
+      // auto-select the first existing item when disabling creation of a new item
+      $scope.modelObject[$scope.modelProperty] = $scope.choices[0][$scope.extract];
+    }
+  });
+
+  // Define a function to show the item value in the select. If no function
+  // is defined, a lookup on a property given by showProperty is done.
+  if (options.showItem) {
```

Comment:
I don't like the `show` naming very much. In angular it's a term often use in relation with visibility. I would advise renaming this to `formatItem`.